python/cudf/cudf/core/dataframe.py

```diff
@@ -1772,9 +1780,12 @@ def as_gpu_matrix(self, columns=None, order="F"):
             raise ValueError("require at least 1 column")
         if nrow < 1:
             raise ValueError("require at least 1 row")
-        dtype = cols[0].dtype
-        if any(dtype != c.dtype for c in cols):
-            raise ValueError("all columns must have the same dtype")
+        if any(
+            (is_categorical_dtype(c) or np.issubdtype(c, np.dtype("object")))
+            for c in cols
+        ):
+            raise TypeError("Data must be homogenous numeric")
```

Comment:
Just a Q: what does "homogeneous numeric" mean in this context?

Reply:
That all columns must be numeric with the same dtype, i.e. not int8 and float32

Reply:
It means that 1. The columns all have to have the same datatype and 2. A piece of me still thinks I'm doing physics :)

Reply:
Although one could argue it's misleading w.r.t the implementation as this code should upcast ints to floats when both are present.

Reply:
@brandon-b-miller I think this could be reworded since it only checks that all of the columns are numerical, not that they're all the same type of numerical. We find a common type and typecast later.